mantidproject/mantid
# Fix issue with beam size in MonteCarloAbsorption

Introduces checks in the main MCAbsorption strategy class to determine whether a generated point will hit the sample or not. It was also discovered that precision issues in the geometry can lead to certain tracks being counted as invalid yet contributing towards the attenuation factor calculation. Invalid tracks are now ignored and regenerated.

See the referenced issue for the problems caused by not checking these issues.

**To test:**

[IndirectCylinderAbsorption.txt](https://github.com/mantidproject/mantid/files/385505/IndirectCylinderAbsorption.txt) is an updated version of the `IndirectCylinderAbsorption.py` algorithm. Download this, rename the extension to `.py` and overwrite the version currently in `PythonInterface/plugins/WorkflowAlgorithms`.

The following script:

``` python
import os

can = Load(Filename='irs26173_graphite002_red.nxs', OutputWorkspace='can')
LoadParameterFile(can, os.path.join(config.getInstrumentDirectory(),'IRIS_Parameters.xml'))
sample = Load(Filename='irs26176_graphite002_red.nxs', OutputWorkspace='sample')
LoadParameterFile(sample, os.path.join(config.getInstrumentDirectory(),'IRIS_Parameters.xml'))

_,factors_no_beam = IndirectCylinderAbsorption(SampleWorkspace='sample', SampleChemicalFormula='H2-O', SampleNumberDensity=0.1,SampleRadius=0.20, 
                                           CanWorkspace='can', UseCanCorrections=True, CanChemicalFormula='V', 
                                           CanRadius=0.25, Events=1000, OutputWorkspace='Cyl_ws-no-beam', CorrectionsWorkspace='Cyl_factors-no-beam',SetBeamSize=False)
plot(factors_no_beam, 0, '-',linestyle='-')
ylim(0,1)

_,factors_beam = IndirectCylinderAbsorption(SampleWorkspace='sample', SampleChemicalFormula='H2-O', SampleNumberDensity=0.1,SampleRadius=0.20, 
                                           CanWorkspace='can', UseCanCorrections=True, CanChemicalFormula='V', 
                                           CanRadius=0.25, Events=1000, OutputWorkspace='Cyl_ws-with-beam', CorrectionsWorkspace='Cyl_factors-with-beam',SetBeamSize=True)
plot(factors_beam, 0, '-',linestyle='-')
ylim(0,1)
```

produces the plots in the referenced issue. After merging these fixes the ratio between the two lines on each curve should be similar.  The data is in the system test data section.

Fixes #17015 

[RELEASE NOTES](https://github.com/mantidproject/mantid/pull/17034/commits/1d21f11da4c505e023c9b5218b3d6e8665d2db9a)

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
